File: frontend/src/lib/api/local-route.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface LocalRouteRule {
  rule_number: number;
  source?: string | null;
  destination?: string | null;
  inbound_interface?: string | null;
  table?: string | null;
  vrf?: string | null;
}

export interface LocalRouteConfigResponse {
  ipv4_rules: LocalRouteRule[];
  ipv6_rules: LocalRouteRule[];
  total_ipv4: number;
  total_ipv6: number;
}

export interface LocalRouteCapabilitiesResponse {
  version: string;
  features: {
    ipv4_local_route: {
      supported: boolean;
      description: string;
    };
    ipv6_local_route: {
      supported: boolean;
      description: string;
    };
    source_matching: {
      supported: boolean;
      description: string;
    };
    destination_matching: {
      supported: boolean;
      description: string;
    };
    inbound_interface_matching: {
      supported: boolean;
      description: string;
    };
    routing_table_selection: {
      supported: boolean;
      description: string;
    };
    vrf_support: {
      supported: boolean;
      description: string;
    };
  };
  device_name?: string;
}

export interface LocalRouteBatchOperation {
  op: string;
  value?: string;
}

export interface LocalRouteBatchRequest {
  rule_number: number;
  rule_type: string;  // 'ipv4' or 'ipv6'
  operations: LocalRouteBatchOperation[];
}

// ============================================================================
// API Service
// ============================================================================

class LocalRouteService {
  /**
   * Get capabilities based on VyOS version
   */
  async getCapabilities(): Promise<LocalRouteCapabilitiesResponse> {
    return apiClient.get<LocalRouteCapabilitiesResponse>("/vyos/local-route/capabilities");
  }

  /**
   * Get all local route configurations
   */
  async getConfig(refresh: boolean = false): Promise<LocalRouteConfigResponse> {
    return apiClient.get<LocalRouteConfigResponse>("/vyos/local-route/config", {
      refresh: refresh.toString(),
    });
  }

  /**
   * Refresh the cached configuration
   */
  async refreshConfig(): Promise<unknown> {
    return apiClient.post("/vyos/config/refresh");
  }

  /**
   * Execute batch operations
   */
  async batchConfigure(request: LocalRouteBatchRequest): Promise<unknown> {
    const result = await apiClient.post("/vyos/local-route/batch", request);
    await this.refreshConfig();
    return result;
  }

  /**
   * Reorder rules
   */
  async reorderRules(
    ruleType: string,
    rules: Array<{ old_number: number; new_number: number; rule_data: LocalRouteRule }>
  ): Promise<unknown> {
    const result = await apiClient.post("/vyos/local-route/reorder", {
      rule_type: ruleType,
      rules,
    });
    await this.refreshConfig();
    return result;
  }

  /**
   * Helper: Create a new rule
   */
  async createRule(
    ruleNumber: number,
    ruleType: string,
    config: Partial<LocalRouteRule>
  ): Promise<unknown> {
    const operations: LocalRouteBatchOperation[] = [];

    // Create rule
    operations.push({
      op: ruleType === "ipv4" ? "set_local_route_rule" : "set_local_route6_rule"
    });

    // Set source
    if (config.source) {
      operations.push({
        op: ruleType === "ipv4" ? "set_local_route_rule_source" : "set_local_route6_rule_source",
        value: config.source
      });
    }

    // Set destination
    if (config.destination) {
      operations.push({
        op: ruleType === "ipv4" ? "set_local_route_rule_destination" : "set_local_route6_rule_destination",
        value: config.destination
      });
    }

    // Set inbound interface
    if (config.inbound_interface) {
      operations.push({
        op: ruleType === "ipv4" ? "set_local_route_rule_inbound_interface" : "set_local_route6_rule_inbound_interface",
        value: config.inbound_interface
      });
    }

    // Set table (required - unless VRF is set)
    if (config.table) {
      operations.push({
        op: ruleType === "ipv4" ? "set_local_route_rule_set_table" : "set_local_route6_rule_set_table",
        value: config.table
      });
    }

    // Set VRF (VyOS 1.5+ only)
    if (config.vrf) {
      operations.push({
        op: ruleType === "ipv4" ? "set_local_route_rule_set_vrf" : "set_local_route6_rule_set_vrf",
        value: config.vrf
      });
    }

    return this.batchConfigure({
      rule_number: ruleNumber,
      rule_type: ruleType,
      operations,
    });
  }

  /**
   * Helper: Update an existing rule
   */
  async updateRule(
    ruleNumber: number,
    ruleType: string,
    config: Partial<LocalRouteRule>
  ): Promise<unknown> {
    const operations: LocalRouteBatchOperation[] = [];

    // IMPORTANT: Process deletes FIRST, then sets
    // This ensures VyOS properly removes fields before setting new values

    // Delete operations first
    if (config.source !== undefined && !config.source) {
      operations.push({
        op: ruleType === "ipv4" ? "delete_local_route_rule_source" : "delete_local_route6_rule_source"
      });
    }

    if (config.destination !== undefined && !config.destination) {
      operations.push({
        op: ruleType === "ipv4" ? "delete_local_route_rule_destination" : "delete_local_route6_rule_destination"
      });
    }

    if (config.inbound_interface !== undefined && !config.inbound_interface) {
      operations.push({
        op: ruleType === "ipv4" ? "delete_local_route_rule_inbound_interface" : "delete_local_route6_rule_inbound_interface"
      });
    }

    if (config.vrf !== undefined && !config.vrf) {
      operations.push({
        op: ruleType === "ipv4" ? "delete_local_route_rule_set_vrf" : "delete_local_route6_rule_set_vrf"
      });
    }

    // Set operations after deletes
    if (config.source !== undefined && config.source) {
      operations.push({
        op: ruleType === "ipv4" ? "set_local_route_rule_source" : "set_local_route6_rule_source",
        value: config.source
      });
    }

    if (config.destination !== undefined && config.destination) {
      operations.push({
        op: ruleType === "ipv4" ? "set_local_route_rule_destination" : "set_local_route6_rule_destination",
        value: config.destination
      });
    }

    if (config.inbound_interface !== undefined && config.inbound_interface) {
      operations.push({
        op: ruleType === "ipv4" ? "set_local_route_rule_inbound_interface" : "set_local_route6_rule_inbound_interface",
        value: config.inbound_interface
      });
    }

    if (config.table !== undefined && config.table) {
      operations.push({
        op: ruleType === "ipv4" ? "set_local_route_rule_set_table" : "set_local_route6_rule_set_table",
        value: config.table
      });
    }

    if (config.vrf !== undefined && config.vrf) {
      operations.push({
        op: ruleType === "ipv4" ? "set_local_route_rule_set_vrf" : "set_local_route6_rule_set_vrf",
        value: config.vrf
      });
    }

    return this.batchConfigure({
      rule_number: ruleNumber,
      rule_type: ruleType,
      operations,
    });
  }

  /**
   * Helper: Delete a rule
   */
  async deleteRule(
    ruleNumber: number,
    ruleType: string
  ): Promise<unknown> {
    const operations: LocalRouteBatchOperation[] = [];
    operations.push({
      op: ruleType === "ipv4" ? "delete_local_route_rule" : "delete_local_route6_rule"
    });

    return this.batchConfigure({
      rule_number: ruleNumber,
      rule_type: ruleType,
      operations,
    });
  }
}

export const localRouteService = new LocalRouteService();

File: backend/routers/local_route/local_route.py
"""
Local Route Router

API endpoints for managing VyOS local route policy configuration.
Supports both IPv4 (local-route) and IPv6 (local-route6) rules.
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import LocalRouteBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
from utils.router_helpers import load_vyos_capabilities
import inspect

router = APIRouter(prefix="/vyos/local-route", tags=["local-route"])

# Stub functions for backwards compatibility with app.py
def set_device_registry(registry):
    """Legacy function - no longer used."""
    pass


def set_configured_device_name(name):
    """Legacy function - no longer used."""
    pass


# ============================================================================
# Pydantic Models
# ============================================================================


class LocalRouteRule(BaseModel):
    """Model for a local route rule (IPv4 or IPv6)."""

    rule_number: int = Field(..., ge=1, le=32765, description="Rule number (1-32765)")
    source: Optional[str] = Field(None, description="Source address or prefix")
    destination: Optional[str] = Field(None, description="Destination address or prefix")
    inbound_interface: Optional[str] = Field(None, description="Inbound interface name")
    table: Optional[str] = Field(None, description="Routing table (1-200 or 'main')")
    vrf: Optional[str] = Field(None, description="VRF instance name (VyOS 1.5+ only)")


class LocalRouteConfigResponse(BaseModel):
    """Response containing all local route rules."""

    ipv4_rules: List[LocalRouteRule] = []
    ipv6_rules: List[LocalRouteRule] = []
    total_ipv4: int = 0
    total_ipv6: int = 0


class LocalRouteCapabilitiesResponse(BaseModel):
    """Response containing feature capabilities."""

    version: str
    features: Dict[str, Any]
    device_name: Optional[str] = None


class LocalRouteBatchOperation(BaseModel):
    """Single operation in a batch request."""

    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value")


class LocalRouteBatchRequest(BaseModel):
    """Model for batch configuration."""

    rule_number: int = Field(..., ge=1, le=32765, description="Rule number")
    rule_type: str = Field(..., description="Rule type: 'ipv4' or 'ipv6'")
    operations: List[LocalRouteBatchOperation]


class LocalRouteReorderRequest(BaseModel):
    """Model for reordering rules."""

    rule_type: str = Field(..., description="Rule type: 'ipv4' or 'ipv6'")
    rules: List[Dict[str, Any]] = Field(
        ..., description="List of rules with old and new numbers"
    )


class VyOSResponse(BaseModel):
    """Standard response from VyOS operations."""

    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities", response_model=LocalRouteCapabilitiesResponse)
async def get_local_route_capabilities(request: Request):
    """
    Get feature capabilities based on device VyOS version.

    Returns feature flags indicating which operations are supported.
    """
    return await load_vyos_capabilities(
        request=request,
        feature_group=FeatureGroup.LOCAL_ROUTE,
        builder_cls=LocalRouteBatchBuilder,
    )


# ============================================================================
# Endpoint 2: Config (Generalized Data)
# ============================================================================


def parse_address_field(value):
    """
    Parse VyOS address field which can be:
    - A string: "10.2.4.5"
    - A dict with 'address' key: {'address': ['10.2.4.5']}
    - None
    """
    if value is None:
        return None
    if isinstance(value, str):
        return value
    if isinstance(value, dict) and "address" in value:
        # VyOS returns {'address': ['10.2.4.5']}
        addresses = value["address"]
        if isinstance(addresses, list) and len(addresses) > 0:
            return addresses[0]
    return None


@router.get("/config", response_model=LocalRouteConfigResponse)
async def get_local_route_config(http_request: Request, refresh: bool = False):
    """
    Get all local route configurations in a generalized format.

    Args:
        refresh: If True, force refresh from VyOS. If False, use cache.

    Returns:
        Generalized configuration data for IPv4 and IPv6 rules
    """
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        ipv4_rules = []
        ipv6_rules = []

        # Parse IPv4 local-route rules
        if "policy" in full_config and "local-route" in full_config["policy"]:
            local_route_config = full_config["policy"]["local-route"]
            if "rule" in local_route_config:
                for rule_num_str, rule_data in local_route_config["rule"].items():
                    rule = LocalRouteRule(
                        rule_number=int(rule_num_str),
                        source=parse_address_field(rule_data.get("source")),
                        destination=parse_address_field(rule_data.get("destination")),
                        inbound_interface=rule_data.get("inbound-interface"),
                        table=rule_data.get("set", {}).get("table") if "set" in rule_data else None,
                        vrf=rule_data.get("set", {}).get("vrf") if "set" in rule_data else None,
                    )
                    ipv4_rules.append(rule)

        # Parse IPv6 local-route6 rules
        if "policy" in full_config and "local-route6" in full_config["policy"]:
            local_route6_config = full_config["policy"]["local-route6"]
            if "rule" in local_route6_config:
                for rule_num_str, rule_data in local_route6_config["rule"].items():
                    rule = LocalRouteRule(
                        rule_number=int(rule_num_str),
                        source=parse_address_field(rule_data.get("source")),
                        destination=parse_address_field(rule_data.get("destination")),
                        inbound_interface=rule_data.get("inbound-interface"),
                        table=rule_data.get("set", {}).get("table") if "set" in rule_data else None,
                        vrf=rule_data.get("set", {}).get("vrf") if "set" in rule_data else None,
                    )
                    ipv6_rules.append(rule)

        # Sort by rule number
        ipv4_rules.sort(key=lambda x: x.rule_number)
        ipv6_rules.sort(key=lambda x: x.rule_number)

        return LocalRouteConfigResponse(
            ipv4_rules=ipv4_rules,
            ipv6_rules=ipv6_rules,
            total_ipv4=len(ipv4_rules),
            total_ipv6=len(ipv6_rules),
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch")
async def local_route_batch_configure(http_request: Request, body: LocalRouteBatchRequest):
    """
    Execute a batch of configuration operations for a local route rule.

    Allows multiple changes in a single VyOS commit for efficiency.
    """
    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = LocalRouteBatchBuilder(version=version)

        # Process operations using inspect for dynamic method calls
        for operation in body.operations:
            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = list(sig.parameters.keys())

            # Build arguments dynamically
            args = []
            if "rule_number" in params:
                args.append(body.rule_number)
            if operation.value and len(params) > len(args):
                args.append(operation.value)

            method(*args)

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Configuration updated"},
            error=response.error if response.error else None,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Endpoint 4: Reorder Rules
# ============================================================================


@router.post("/reorder")
async def local_route_reorder_rules(http_request: Request, body: LocalRouteReorderRequest):
    """
    Reorder local route rules by renumbering them.

    This is done by deleting all rules and recreating them with new numbers
    in a single batch operation.
    """
    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = LocalRouteBatchBuilder(version=version)

        # Determine operation prefix based on rule type
        if body.rule_type == "ipv4":
            delete_op = "delete_local_route_rule"
            create_op = "set_local_route_rule"
            source_op = "set_local_route_rule_source"
            dest_op = "set_local_route_rule_destination"
            interface_op = "set_local_route_rule_inbound_interface"
            table_op = "set_local_route_rule_set_table"
            vrf_op = "set_local_route_rule_set_vrf"
        else:  # ipv6
            delete_op = "delete_local_route6_rule"
            create_op = "set_local_route6_rule"
            source_op = "set_local_route6_rule_source"
            dest_op = "set_local_route6_rule_destination"
            interface_op = "set_local_route6_rule_inbound_interface"
            table_op = "set_local_route6_rule_set_table"
            vrf_op = "set_local_route6_rule_set_vrf"

        # Step 1: Delete all rules in reverse order
        old_numbers = [rule["old_number"] for rule in body.rules]
        for old_num in sorted(old_numbers, reverse=True):
            getattr(builder, delete_op)(old_num)

        # Step 2: Recreate rules with new numbers
        for rule in body.rules:
            new_num = rule["new_number"]
            rule_data = rule["rule_data"]

            # Create rule
            getattr(builder, create_op)(new_num)

            # Set properties
            if rule_data.get("source"):
                getattr(builder, source_op)(new_num, rule_data["source"])
            if rule_data.get("destination"):
                getattr(builder, dest_op)(new_num, rule_data["destination"])
            if rule_data.get("inbound_interface"):
                getattr(builder, interface_op)(new_num, rule_data["inbound_interface"])
            if rule_data.get("table"):
                getattr(builder, table_op)(new_num, rule_data["table"])
            if rule_data.get("vrf"):
                getattr(builder, vrf_op)(new_num, rule_data["vrf"])

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Rules reordered successfully"},
            error=response.error if response.error else None,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
